crates/adapters/src/transport/pubsub/input.rs

```diff
@@ -0,0 +1,246 @@
+use crate::{
+    transport::Step, InputConsumer, InputEndpoint, InputReader, PipelineState,
+    TransportInputEndpoint,
+};
+use anyhow::{anyhow, bail, Error as AnyError, Result as AnyResult};
+use chrono::DateTime;
+use dbsp::circuit::tokio::TOKIO;
+use futures::StreamExt;
+use google_cloud_gax::conn::Environment;
+use google_cloud_pubsub::{
+    client::{google_cloud_auth::credentials::CredentialsFile, Client, ClientConfig},
+    subscription::{SeekTo, Subscription},
+};
+use log::{debug, warn};
+use pipeline_types::{program_schema::Relation, transport::pubsub::PubSubInputConfig};
+use std::{
+    sync::Arc,
+    thread,
+    time::{Duration, SystemTime},
+};
+use tokio::{
+    sync::watch::{channel, Receiver, Sender},
+    task::JoinHandle,
+};
+use tokio_util::sync::CancellationToken;
+
+pub struct PubSubInputEndpoint {
+    config: Arc<PubSubInputConfig>,
+}
+
+impl PubSubInputEndpoint {
+    pub fn new(config: PubSubInputConfig) -> Result<Self, AnyError> {
+        // Validation.
+        if config.snapshot.is_some() && config.timestamp.is_some() {
+            bail!("'snapshot' and 'timestamp' config options cannot be set at the same time",);
```

Comment:
In other places this used to be a Result

Reply:
This will return a Result